### PR TITLE
feat: add jitter

### PR DESCRIPTION
**Pull request by s0nicboOm**
1. Add jitter
      -  for caches
      -  for model saving
2. Support RegistryConf and JitterConf

### CODE REVIEW

File: numalogic/registry/redis_registry.py
Comment (qhuai) [inline numalogic/registry/redis_registry.py:216]: optionally, we can have a helper function such as ```value = apply_jitter(value, jitter, step)```.

related with my earlier comments, if jitter_steps is truly in minute unit, suggest to name the variable as ```jitter_step_minutes``` which makes the difference with ```jitter_secs``` immediately clear.

Comment (s0nicboOm) [reply numalogic/registry/redis_registry.py:216]: Yes. It should be in minutes . That is a fair ask. Let me change the variable name to have minute info.

File: numalogic/registry/artifact.py
Comment (s0nicboOm) [reply numalogic/registry/artifact.py:150]: 2. Do you see an issue with the current way? Ideally ttl, jitter_sec value that is set should be such that ttl>>>jitter_sec. What do you think? That `abs` is just a sanity check.
3. Jitter steps should be in minutes. Let me fix that and make it more intuitive with the variable name.

Comment (s0nicboOm) [reply numalogic/registry/artifact.py:150]: Regarding (1), I want to have steps as mins. I have fixed that. Thanks for pointing that out

File: numalogic/registry/redis_registry.py
Comment (s0nicboOm) [reply numalogic/registry/redis_registry.py:65]: Addressed in the comments.
This scenario won't hapen.

File: numalogic/registry/artifact.py
Comment (s0nicboOm) [reply numalogic/registry/artifact.py]: Regarding (2), Have added a check for the positive vale in the condition itslef.

File: numalogic/config/_config.py
Comment (qhuai) [inline numalogic/config/_config.py:38]: Nitpick: a generic JitterConf doesn't need to hardcode a (physical) dimension. This actually is a TimeJitterConf.

Another nitpick: jitter_secs is plural, jitter_steps_min is singular, model_expiry_sec is singular, retry_secs is plural.
Again, it's a nitpick so no need to change those terms in this PR. but it's one aspect to pay attention to in future PRs.

File: numalogic/registry/artifact.py
Comment (qhuai) [inline numalogic/registry/artifact.py:133]: Is this `min()` redundant?

`min(ts - jitter_secs, ts)` always equals to `ts - jitter_secs` unless jitter_secs is negative or the minus operation underflows.

**Review by qhuai — APPROVED** [review]
Have a couple of minor comments. Overall it looks good to me.

File: numalogic/config/_config.py
Comment (s0nicboOm) [reply numalogic/config/_config.py:38]: Let me add a TODO for this.

File: numalogic/registry/artifact.py
Comment (s0nicboOm) [reply numalogic/registry/artifact.py:133]: aaahhh....This should have been `max` and not `min`.

File: numalogic/registry/artifact.py
Comment (qhuai) [reply numalogic/registry/artifact.py:132]: Thinking more on this, I think a generic alternative could be the following. That being said, I will leave the final call to you.

```
_apply_jitter(min_val: int, max_val: int, bins: int = 100):
  delta = max_val - min_val
  assert(0 < bins and bins < delta)
  return random.randrange(min_val, max_val, delta / bins)
```

Comment (s0nicboOm) [reply numalogic/registry/artifact.py]: That change was still in works as I wanted to make it more robust :) .  Yes you are 100% correct on that.

These are the final changes I was looking for:
```
if jitter_sec == jitter_steps_min == 0:
        return ts
    if jitter_sec < 60 * jitter_steps_min:
        raise ConfigError("jitter_sec should be at least 60*jitter_steps_min")
    begin = ts if ts - jitter_sec < 0 else ts - jitter_sec
    end = ts + jitter_sec + 1
    return random.randrange(begin, end, 60 * jitter_steps_min)
```
a) The specific reason I do not want to keep it zero is because it defeats the purpose of TTL. The function should return a **Positive** value always. If random function returns 0 that means the TTL for storing becomes zero. That is misleading.  The above code will take care of that.

b) Let's say we pass a config such that ts = 1000, jitter_sec= 5 sec, and jitter_steps_min=2
In this scenario we will always return ts =1.  (always). This is the scenario you are referring to.  We have two options here. Either we throw a Configuration Error or we ignore it and jitter . Even though function will return the same value, it defeats the purpose of Jitter Conf. I have added a config check for that.
Here, we need to check : `(jitter_sec * 2) // (60 * jitter_steps_min) > 1` and throw config error.

c) There is another scenario such that ts = 1, jitter_sec= 5 sec, and jitter_steps_min=2.
In this case, the condition that should hold true is: "jitter_sec //(60 * jitter_steps_min)> 1"

From (2) and (3), if we union the conditions (3) holds true and hence you see the above condition. 

What do you think?

Comment (qhuai) [reply numalogic/registry/artifact.py]: The logic looks good.

Don't mean to prolong the completion of this PR but when you need to multiple a hardcode 60x to the jitter steps at every use of the variable. It may be simpler to keep the jitter range and step in the same (physical) dimension and unit. Again, it's something optional to consider.

Comment (s0nicboOm) [reply numalogic/registry/artifact.py]: That is a fair ask.

No worries! I think I will have to wait till Monday for other's review. anyways :)

**Review by ab93 — APPROVED** [review]
Minor comment

File: numalogic/registry/redis_registry.py
Comment (s0nicboOm) [reply numalogic/registry/redis_registry.py]: fixed

File: numalogic/config/_config.py
Comment (s0nicboOm) [reply numalogic/config/_config.py]: It has been updated to jitter_steps_min. It is the min differnce between two random value that the function can return. In other words, the interval gap between two jitter values.

**Pull request merged**
